rules/no-error-ctor-with-notthrows.js

```diff
@@ -0,0 +1,50 @@
+'use strict';
+const {visitIf} = require('enhance-visitors');
+const util = require('../util');
+const createAvaRule = require('../create-ava-rule');
+
+const errorNameRegex = /^([A-Z][a-z\d]*)*Error$/;
+
+const create = context => {
+	const ava = createAvaRule();
+
+	return ava.merge({
+		CallExpression: visitIf([
+			ava.isInTestFile,
+			ava.isInTestNode
+		])(node => {
+
+			if (typeof node.callee.property === 'undefined') {
+				return;
+			}
+
+			const functionArgIndex = node.arguments.length - 1;
+
+			if (functionArgIndex !== 1) {
+				return;
+			}
+
+			const calleeProperty = node.callee.property.name;
+			const functionArgName = node.arguments[1].name;
+
+			if (calleeProperty === 'notThrows' || calleeProperty === 'notThrowsAsync') {
+				if (errorNameRegex.test(functionArgName)) {
```

Comment:
I think we can simplify this to 
```js
functionArgName.endsWith("Error")
```

`errorNameRegex` implies that a valid identifier name will be of the format `^([A-Z][a-z\d]*)*`. However it is not sufficient, per [ECMAScript](https://www.ecma-international.org/ecma-262/#prod-IdentifierName), the following identifier name is also valid
```js
var $DOMError;
var \u{1F431}Error;
```
Instead of we match the identifierName via syntax, we can simply test if it ends up with `Error`.

Reply:
For cases like,
```
valid: const test = require('ava');
  
  		test('some test',t => {
  			t.notThrows(() => {
  				t.pass();
  			}, {firstName:'some', lastName: 'object'});
  		});
```
it will throw error, as `functionArgName` is `undefined` here and regex method won't fail in this case as it results into false.

How about keeping regex here and also checking if `functionArgName` is not undefined ( like you mentioned above ), would be both efficient and less error-prone?

Reply:
I think regex would miss some use cases when user has different name style than CamelCase. Anyway I have also constructed exotic error class name that the regex would not throw, i.e. `$DOMError` and `\u{1F431}Error`.

Per [ESTree spec](https://github.com/estree/estree/blob/master/es5.md), only `Identifier` node will have `name` property, so I think we can check if `name` is defined, which is equivalent to check if `node` is Identifier.

In all, my advice on the checking process will be:

1) check AST pattern on `t.notThrows` and `t.throws` call expression. Skip on unmatched AST shapes.
2) check their signatures. Skip on invalid ones.
3) test if `functionArgName ` ends with error, which I think is simpler, or test if it complies to a [complex identifier name regex](https://github.com/babel/babel/blob/v7.5.4/packages/babel-parser/src/util/identifier.js#L71) for spec compliance. Throw if the test returns true.

Reply:
Yes, let's go with it.